executor: load data/batch insert improvement reducing memory allocation/free

cfzjywxk opened the pull request:
<!--
Thank you for contributing to TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

### What problem does this PR solve? <!--add issue link with summary if exists-->

- reducing the memory allocation/free and copy times during batch insert/load data execution process, trying to improve the performance
- refactor some code logic

**plz help taking a look for row buffers if they are still used after one batch commitment, this pr will make these memories reused after one batch transaction committing, by now unit tests show good.**


### What is changed and how it works?

do not alloc mem for every row per batch, use one mem buf for all batches

### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

 - Unit test


Code changes

 - Has exported function/method change
 - Has exported variable/fields change


Side effects

 - Possible performance regression
 - Increased code complexity

jackysp commented:
Could you provide some memory usage test results or performance test results?

cfzjywxk commented:
> Could you provide some memory usage test results or performance test results?
@jackysp 
500w data load
seems not much difference on time cost and memory consume...
need to take a closer look...

master branch process memory usage
VmPeak:  6419928 kB
VmSize:  6419928 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:   2044400 kB
VmRSS:   2044400 kB
VmData:  2644572 kB
VmStk:       132 kB
VmExe:     44828 kB
VmLib:      1532 kB
VmPTE:      4524 kB
VmPMD:        36 kB
VmSwap:        0 kB


dev branch process memory usage
VmPeak:  6219084 kB                                                      
VmSize:  6219084 kB                                                      
VmLck:         0 kB                                                      
VmPin:         0 kB                                                      
VmHWM:   1729308 kB                                                      
VmRSS:   1729308 kB                                                      
VmData:  2378320 kB                                                      
VmStk:       132 kB                                                      
VmExe:     44828 kB                                                      
VmLib:      1532 kB                                                      
VmPTE:      3916 kB 

virtual mem and residul not diff much

jackysp commented:
/run-all-tests

cfzjywxk commented:
@jackysp @lysu  any suggestions ? 
eh.. not as expected ..
old implementation will "make as much as  input rows mem buf" and for each row there will be a copy between "filled row and row buf",  but the results seems not differ too much. 

maybe Golang did a real good mem cache job ...

jackysp commented:
I think it should work, we can try it on the IDC server.

lysu commented:
> @jackysp @lysu any suggestions ?
> eh.. not as expected ..
> old implementation will "make as much as input rows mem buf" and for each row there will be a copy between "filled row and row buf", but the results seems not differ too much.
> 
> maybe Golang did a real good mem cache job ...

Golang will managed memory, so view rss in os level maybe useless - -, maybe can take look go's memstat https://blog.xiaoba.me/2017/09/02/how-to-play-golang-with-gdb-and-python.html.. but IMHO it's hard to observe too - - maybe take look other metric like import speed? :D

cfzjywxk commented:
500w data load this is tested on idc machine， speed and process mem use almost the same(not very stable, sometimes even slower)

lysu commented on executor/load_data.go:
```diff
@@ -112,7 +112,12 @@ type LoadDataInfo struct {
+	if uint64(cap(e.rows)) < limit {
+		e.rows = make([][]types.Datum, 0, limit)
+		for i := 0; uint64(i) < limit; i++ {
+			e.rows = append(e.rows, make([]types.Datum, len(e.Table.Cols())))
```
maybe we only need new `limit - cap(e.rows)` types.Datum here if the limit can be dynamical change 

and the other optimize maybe or not work is `make([]types.Datum, (limit -cap(e.rows)) * len(e.Table.Cols()))` once and then cut slice from this big array to append

cfzjywxk replied:
seems trace not support load data now ?

cfzjywxk replied:
@lysu currently init will be called only once. in another pr making variable settable will do this

ngaut commented:
Please squash your commit message before you file the PR.

cfzjywxk commented:
> I think it should work, we can try it on the IDC server.

retest on idc server

500w rows load data into empty table


```
dev branch:

time cost(twice)
mysql> load data local infile '/data3/rwork/tests/data/500w.dat' into table p3438_2;
Query OK, 5000000 rows affected (10 min 42.26 sec)
Records: 5000000  Deleted: 0  Skipped: 0  Warnings: 0
Query OK, 5000000 rows affected (13 min 17.68 sec)
Records: 5000000  Deleted: 0  Skipped: 0  Warnings: 0

memory usage from golang pprof
(pprof) top
Showing nodes accounting for 100.29MB, 57.58% of 174.19MB total
Dropped 216 nodes (cum <= 0.87MB)
Showing top 10 nodes out of 219
      flat  flat%   sum%        cum   cum%
   54.65MB 31.37% 31.37%    67.55MB 38.78%  compress/flate.NewWriter
    7.50MB  4.31% 35.68%     7.50MB  4.31%  github.com/pingcap/tidb/util/chunk.newFixedLenColumn
    6.95MB  3.99% 39.67%     6.95MB  3.99%  compress/flate.(*compressor).initDeflate (inline)
    6.91MB  3.97% 43.64%     6.91MB  3.97%  google.golang.org/grpc/transport.newBufWriter
       5MB  2.87% 46.51%        5MB  2.87%  github.com/prometheus/common/expfmt.labelPairsToText
    4.50MB  2.58% 49.09%    61.14MB 35.10%  github.com/prometheus/common/expfmt.MetricFamilyToText
    4.06MB  2.33% 51.43%     4.06MB  2.33%  bytes.makeSlice
    3.71MB  2.13% 53.56%    12.90MB  7.41%  compress/flate.(*compressor).init
    3.50MB  2.01% 55.57%     3.50MB  2.01%  github.com/pingcap/parser.yyParse
    3.50MB  2.01% 57.58%     3.50MB  2.01%  github.com/prometheus/client_golang/prometheus.(*histogram).Write
```

```
master branch
time cost(once)
mysql> load data local infile '/data3/rwork/tests/data/500w.dat' into table master_2;
^[[A^[[BQuery OK, 5000000 rows affected (18 min 33.82 sec)
Records: 5000000  Deleted: 0  Skipped: 0  Warnings: 0

memory usage from golang pprof

Showing nodes accounting for 216.89MB, 49.27% of 440.20MB total
Dropped 338 nodes (cum <= 2.20MB)
Showing top 10 nodes out of 218
      flat  flat%   sum%        cum   cum%
  120.76MB 27.43% 27.43%   155.05MB 35.22%  compress/flate.NewWriter
   24.51MB  5.57% 33.00%    24.51MB  5.57%  github.com/pingcap/tidb/util/chunk.newFixedLenColumn
   16.01MB  3.64% 36.64%    16.01MB  3.64%  compress/flate.(*compressor).initDeflate (inline)
   10.59MB  2.41% 39.04%    34.29MB  7.79%  compress/flate.(*compressor).init
   10.50MB  2.39% 41.43%    10.50MB  2.39%  github.com/prometheus/client_golang/prometheus.(*histogram).Write
       8MB  1.82% 43.25%   134.85MB 30.63%  github.com/prometheus/common/expfmt.MetricFamilyToText
       7MB  1.59% 44.84%        7MB  1.59%  github.com/prometheus/common/expfmt.labelPairsToText
    6.52MB  1.48% 46.32%    20.53MB  4.66%  github.com/prometheus/client_golang/prometheus.(*Registry).Gather
    6.50MB  1.48% 47.79%     6.50MB  1.48%  github.com/pingcap/tidb/executor.schema2ResultFields
    6.50MB  1.48% 49.27%    23.50MB  5.34%  google.golang.org/grpc.newClientStream
```

time cost is not very stable,  get alloc_space metrict from pprof after loading

cfzjywxk commented:
@jackysp PTAL

jackysp commented:
On IDC, we can use ansible to start tikv as the storage engine. You could take a look at the tidb heap separately during the test. @lysu will help.

cfzjywxk commented:
some tests results on idc tidb cluster,  loading 500w rows sysbench table schema 

 

```

master branch

Git Commit Hash: fcc15b1256e4e33132a80e39b841a6e062c353a0

mysql> load data local infile '/data3/rwork/tests/data/500w.dat' into table 2_master;
Query OK, 5000000 rows affected (13 min 39.47 sec)
Records: 5000000 Deleted: 0 Skipped: 0 Warnings: 0

Type: alloc_space
Time: Jul 18, 2019 at 7:34pm (CST)
Showing nodes accounting for 101682.08MB, 96.52% of 105347.99MB total
Dropped 775 nodes (cum <= 526.74MB)
flat flat% sum% cum cum%
42212.87MB 40.07% 40.07% 42212.87MB 40.07% github.com/pingcap/tidb/executor.(*LoadDataInfo).getLine
21196.27MB 20.12% 60.19% 21196.27MB 20.12% github.com/pingcap/goleveldb/leveldb/memdb.New
10225.16MB 9.71% 69.90% 10225.16MB 9.71% github.com/pingcap/goleveldb/leveldb/memdb.(*DB).Put
3374.63MB 3.20% 73.10% 3374.63MB 3.20% github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).initKeysAndMutations.func1
2650.29MB 2.52% 75.62% 2650.29MB 2.52% github.com/pingcap/tidb/store/tikv.(*RegionCache).GroupKeysByRegion

```

 

```

dev branch(master commit position with  dev commits)

Git Commit Hash: 7f6f89c80f39a5a6e0022fd79a9b9af85fadfff9

mysql> load data local infile '/data3/rwork/tests/data/500w.dat' into table p4348_2;
Query OK, 5000000 rows affected (10 min 42.53 sec)
Records: 5000000 Deleted: 0 Skipped: 0 Warnings: 0

 

Build ID: b6b3aac95012ab192d0a3e1110aebf241dd40a9f
Type: alloc_space
Time: Jul 18, 2019 at 7:52pm (CST)
Showing nodes accounting for 101016.81MB, 96.95% of 104191.50MB total

41964.72MB 40.28% 40.28% 41964.72MB 40.28% github.com/pingcap/tidb/executor.(*LoadDataInfo).getLine
21250.88MB 20.40% 60.67% 21250.88MB 20.40% github.com/pingcap/goleveldb/leveldb/memdb.New
10198.45MB 9.79% 70.46% 10198.45MB 9.79% github.com/pingcap/goleveldb/leveldb/memdb.(*DB).Put
3477.13MB 3.34% 73.80% 3477.13MB 3.34% github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).initKeysAndMutations.func1
2684.87MB 2.58% 76.37% 2685.37MB 2.58% github.com/pingcap/tidb/store/tikv.(*RegionCache).GroupKeysByRegion
2312.38MB 2.22% 78.59% 2312.38MB 2.22% github.com/pingcap/tidb/util/codec.reallocBytes
1815.13MB 1.74% 80.34% 1815.13MB 1.74% github.com/pingcap/tidb/executor.(*fieldWriter).outputField

```

ngaut commented:
According to the memory profile output, there is not much difference between master and current PR.
It's wired, since the loading time reduce 25%.

cfzjywxk commented:
thanks @lysu  for another mem usage opt suggestion for "getLine" function, 
rerun with update, 

memory usage better:
Showing nodes accounting for 59771.25MB, 95.65% of 62486.68MB total
Dropped 665 nodes (cum <= 312.43MB)
      flat  flat%   sum%        cum   cum%
21339.26MB 34.15% 34.15% 21339.26MB 34.15%  github.com/pingcap/goleveldb/leveldb/memdb.New
10320.54MB 16.52% 50.67% 10320.54MB 16.52%  github.com/pingcap/goleveldb/leveldb/memdb.(*DB).Put

load data local infile '/data3/rwork/tests/data/500w.dat' into table p4348_3;
Query OK, 5000000 rows affected (11 min 22.50 sec)
Records: 5000000  Deleted: 0  Skipped: 0  Warnings: 0


rt higger coz pd machine has some noise... (maybe we need more dev/test machine to deploy clean test env)
[pd.go:109] ["get timestamp too slow"]
need further tests

cfzjywxk commented:
@jackysp @lysu  should we merge this first and leave other optimizations in other prs so that this will not be too big diff ?

jackysp commented on executor/load_data.go:
```diff
@@ -16,6 +16,7 @@ package executor
 import (
 	"context"
 	"fmt"
+	"github.com/pingcap/tidb/util/hack"
```
Please move it to the 3rd party libs.

jackysp commented on executor/builder.go:
```diff
@@ -681,6 +681,9 @@ func (b *executorBuilder) buildLoadData(v *plannercore.LoadData) Executor {
 		},
 	}
 
+	var defaultLoadDataBatchCnt uint64 = 20000 //TODO this will be changed to variable in another pr
```
Comment should start with whitespace.

cfzjywxk commented:
/run-all-tests

cfzjywxk commented:
/run-all-tests

jackysp commented:
PTAL @lysu

cfzjywxk commented:
/run-circleci-tests
